sycl/doc/extensions/InvokeSIMD/SYCL_INTEL_invoke_simd.asciidoc

```diff
@@ -0,0 +1,364 @@
++sycl::ext::oneapi::uniform+.  Any such pointer value must point to memory
+that is accessible by all work-items in the sub-group (i.e. the pointer must
+point to an allocation in local or global memory).  The address space for such
+pointers can be local, global or generic; if a generic pointer that points
+to an allocation in private memory is passed as an argument, the behavior is
+undefined.
+
+In order to invoke the SIMD function, +invoke_simd+ converts each argument
+in the +args+ parameter pack according to the following rules:
+
+- Arguments of type +bool+ are converted to type
+  +sycl::ext::oneapi::simd_mask<bool, N>+, where +N+ is the sub-group size of
+  the calling kernel.  Element +i+ of the SIMD type represents the value from
+  the work-item with sub-group local ID +i+.
+
+- All other arithmetic arguments of type +T+ are converted to type
+  +sycl::ext::oneapi::simd<T, N>+, where +N+ is the sub-group size of the
+  calling kernel.  Element +i+ of the SIMD type represents the value from the
+  work-item with sub-group local ID +i+.
+
+- Arguments of type +sycl::ext::oneapi::uniform<T>+ are converted to type +T+.
+  The value may be taken from any work-item in the sub-group; wrapping an
+  argument in a +sycl::ext::oneapi::uniform+ is an assertion that the variable
+  holds the same value for all work-items in the sub-group. If this assertion
+  does not hold, the value of the scalar variable passed to the SIMD function
+  is undefined.
+
+- Arguments of type +std::tuple+ are mapped as if each member of the tuple
+  was mapped individually (e.g. a +std::tuple<float, int>+ becomes a
+  +std::tuple<sycl::ext::oneapi::simd<float, N>, sycl::ext::oneapi::simd<int, N>>+.
+
+After this mapping has taken place, an appropriate variant of +Callable+ is
+selected based on standard overload resolution.
+
+The return value of an +invoke_simd+ function is subject to the same mapping
+rules in reverse:
+
+- Return values of type +sycl::ext::oneapi::simd_mask<bool, N>+ are converted
+  to +bool+, and the value in element +i+ of the SIMD type is returned to the
+  work-item with sub-group local ID +i+.
+
+- Return values of all other arithmetic types +sycl::ext::oneapi::simd<T, N>+
+  are converted to +T+, and the value in element +i+ of the SIMD type is
+  returned to the work-item with sub-group local ID +i+.
+
+- Return values of type +T+ are converted to +sycl::ext::oneapi::uniform<T>+,
+  and broadcast to each work-item; every work-item in the sub-group receives
+  the same value.
+
+- Return values of type +std::tuple+ are mapped as if each memebr of the tuple
+  was mapped individually (e.g. a +std::tuple<sycl::ext::oneapi::simd<float, N>, sycl::ext::oneapi::simd<int, N>>+
+  becomes a +std::tuple<float, int>+).
+
+Note that these conversion rules do not permit user-defined types or special
+types (e.g. +nd_item+) being passed between SPMD and SIMD contexts.  This
```

Comment:
I'm not sure that this is fully enforced, because as stated above:
> Each argument in the +args+ parameter pack must be an arithmetic type, a
> trivially copyable type wrapped in a +sycl::ext::oneapi::uniform+, or a
> +std::tuple+ of these types.

It is possible to pass a user-defined trivially copyable type to `uniform` wrapper and therefor make it cross the border between those two contexts

Reply:
Good catch.  I think this should say something like the below, instead:
> Note that special SYCL types (e.g. +nd_item+) cannot be passed between SPMD and SIMD contexts.

@rolandschulz, do you agree?  Should we go one step further and disallow the creation of `uniform<T>` objects if `T` is a special type?

Reply:
I think we'd have to carefully spell out what special types we're talking about, but it makes sense that something that specifically is intended to vary across work-items shouldn't be able to passed to uniform. 

I would hope that the compiler would scream about this anyway, but making it formally an error is probably more responsible. It's interesting to think of what this means in the context of SG size 1, but I don't think that's a special case worth making exceptions for.

Reply:
> I think we'd have to carefully spell out what special types we're talking about, but it makes sense that something that specifically is intended to vary across work-items shouldn't be able to passed to uniform.

Agreed.  Added this in c1a865ef8d66.

> It's interesting to think of what this means in the context of SG size 1, but I don't think that's a special case worth making exceptions for.

Yeah.  We also can't detect whether the sub-group size is 1 or not at (host) compile-time, so it could be difficult to special case.